Sprintfinity3902/HudMenu/DungeonHud.cs

```diff
@@ -2,33 +2,150 @@
+        private IRoom CurrentRoom;
+        //public List<IEntity> Icons { get; private set; }
+        public List<IEntity> Map { get; private set; }
+        public List<int> AlreadyInMap { get; set; }
+        public IDungeon Dungeon { get; protected set; }
 
-        public DungeonHud(Game1 game)
+        public DungeonHud(Game1 game, IDungeon dungeon)
         {
             Game = game;
             Link = Game.playerCharacter;
+            MapPickup = false;
             Icons = new List<IEntity>();
-            WorldPoint = new Vector2(0, -88 * Global.Var.SCALE);
+            Map = new List<IEntity>();
+            AlreadyInMap = new List<int>();
+            hudInitializer = new HudInitializer(this);
+            Dungeon = dungeon;
+            RoomLocations = dungeon.RoomLocations;
+            CurrentRoom = dungeon.CurrentRoom;
+            WorldPoint = new Vector2(0, 0);
 
             Initialize();
         }
 
+        public override void Initialize()
+        {
+            hudInitializer.InitializeDungeonHud(RoomLocations);
+        }
+
         public override void Update(GameTime gameTime)
         {
 
+            /*Add every room Link is in into the Map List*/
+            if (!AlreadyInMap.Contains(Dungeon.CurrentRoom.Id))
+            {
+                Map.Add(GetNewRoom(Dungeon.CurrentRoom));
+                AlreadyInMap.Add(Dungeon.CurrentRoom.Id);
+                /*If map has already been picked up, add to Icons as well*/
+                if (MapPickup) Icons.Add(GetNewRoom(Dungeon.CurrentRoom));
+            }
+
+            /*This will build the initial map using the rooms Link has already been in when map is picked up*/
+            if (!MapPickup && Link.itemcount[IItem.ITEMS.MAP] > 0) 
+            {
+                int x = ((Dungeon.CurrentRoom.RoomPos.X -1) * BLOCK_SIZE + 2) * Global.Var.SCALE;
+                int y = ((Dungeon.CurrentRoom.RoomPos.Y) * BLOCK_SIZE + 2) * Global.Var.SCALE;
+                LinkBlock = new YellowLinkBlock(new Vector2( x + DUNGEON_INSIDE_MAP_X, y + DUNGEON_INSIDE_MAP_Y));
+                foreach (IEntity room in Map)
+                {
+                    Icons.Add(room);
+                }
+                Icons.Add(LinkBlock);
+                MapPickup = true;
+            }
+
+            if (MapPickup && (CurrentRoom.Id != Dungeon.CurrentRoom.Id) )
+            {
+                Icons.Remove(LinkBlock);
+                int deltaX = (Dungeon.CurrentRoom.RoomPos.X - CurrentRoom.RoomPos.X) * BLOCK_SIZE * Global.Var.SCALE;
+                if (deltaX != 0)
+                {
+                    LinkBlock.X += deltaX;
+                }
+                else
+                {
+                    int deltaY = (Dungeon.CurrentRoom.RoomPos.Y - CurrentRoom.RoomPos.Y) * BLOCK_SIZE * Global.Var.SCALE;
+                    LinkBlock.Y += deltaY;
+                }
+                Icons.Add(LinkBlock);
+                CurrentRoom = Dungeon.CurrentRoom;
+            }
```

Comment:
This is alot to have in a update function. Does this all have to be completed every frame? If it does, this is at a minimum should be placed in a function that is called by the update function.

Reply:
I think all of these if statements have a more optimal solution.

IDEALLY:

- Map checks if a new room needs to be added each time you enter a new room
- Similar thought behind drawing link's current location ^
- The item pickup tells the hud that it has been picked up, so the hud doesn't have to continually check

I wouldn't say this is required for this PR, but should be kept in our thoughts for future pr's. Adding functions to the DungeonHud should be added this PR though.